driver/kvm: update libkvmi to v6

adlazar opened the pull request:
optimize on receive and send messages

    use sendmsg() instead of send()
    add iobuf when reading from socket

add some functions

    kvmi_memory_mapping()
    kvmi_check_command()
    kvmi_check_event()
    kvmi_get_mtrr_type()

remove vcpu parameter from some commands

    set_page_access, set_page_write_bitmap ...

replace kvmi_inject_page_fault() with kvmi_inject_exception()

    use vector=14

add batch messages
    Send multiple messages with one syscall.

    Eg.
    KVMI_SET_PAGE_ACCESS(oneway)+KVMI_SET_REGISTERS(oneway)+KVMI_EVENT_REPLY

use timeouts on all socket operations

accept command replies from older/newer versions of KVMI (kernel side)

replaced MAP_SHARED with MAP_PRIVATE when using remote mapping

    There is no need to share the mappings created via introspection
    engine requests.

Signed-off-by: Cătălin Marian Pichiu <cpichiu@bitdefender.com>
Signed-off-by: Mihai Donțu <mdontu@bitdefender.com>
Signed-off-by: Nicușor Cîțu <ncitu@bitdefender.com>
Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

adlazar committed:
driver/kvm: update libkvmi to v6

optimize on receive and send messages

    use sendmsg() instead of send()
    add iobuf when reading from socket

add some functions

    kvmi_memory_mapping()
    kvmi_check_command()
    kvmi_check_event()
    kvmi_get_mtrr_type()

remove vcpu parameter from some commands

    set_page_access, set_page_write_bitmap ...

replace kvmi_inject_page_fault() with kvmi_inject_exception()

    use vector=14

add batch messages
    Send multiple messages with one syscall.

    Eg.
    KVMI_SET_PAGE_ACCESS(oneway)+KVMI_SET_REGISTERS(oneway)+KVMI_EVENT_REPLY

use timeouts on all socket operations

accept command replies from older/newer versions of KVMI (kernel side)

replaced MAP_SHARED with MAP_PRIVATE when using remote mapping

    There is no need to share the mappings created via introspection
    engine requests.

Signed-off-by: Cătălin Marian Pichiu <cpichiu@bitdefender.com>
Signed-off-by: Mihai Donțu <mdontu@bitdefender.com>
Signed-off-by: Nicușor Cîțu <ncitu@bitdefender.com>
Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

Commit:
driver/kvm/libkvmi: add kvmi_get_maximum_gfn()

Signed-off-by: Ștefan Sicleru <ssicleru@bitdefender.com>

Wenzel commented on libvmi/driver/kvm/include/kvmi/libkvmi.h:
```diff
@@ -102,10 +111,17 @@ int     kvmi_get_registers( void *dom, unsigned short vcpu, struct kvm_regs *reg
+int     kvmi_queue_registers( void *batch, unsigned short vcpu, const struct kvm_regs *regs );
+int     kvmi_queue_reply_event( void *batch, unsigned int msg_seq, const void *data, size_t data_size );
+int     kvmi_queue_page_access( void *batch, unsigned long long int *gpa, unsigned char *access, unsigned short count );
+int     kvmi_queue_pause_vcpu( void *batch, unsigned short vcpu );
```
@adlazar i believe there is a mismatch between the `libkvmi.h` v6 from KVM and this one in the PR,
because the `kvmi_get_maximum_gf` API is missing here.
https://github.com/adlazar/kvm/blob/kvmi-v6/tools/kvm/kvmi/include/kvmi/libkvmi.h#L125

Wenzel replied:
I just ran a `diff` and it's the only API missing.

mtarral replied:
The implementation is missing too.

adlazar replied:
Commit c927ebb adds the new command.

adlazar committed:
driver/kvm: update kvm_pause_vm() to KVMI v6

The former version of kvmi_pause_all_vcpus() returned the online vCPU
count from kernel (those which will send the pause events).

The new implementation (from libkvmi.c) is a simple for loop which
needs the number of online vCPUs. If it make sense, we can change
kvmi_pause_all_vcpus() later to get that number from kernel.

Signed-off-by: Adalbert Lazăr <alazar@bitdefender.com>

Wenzel committed:
Merge branch 'kvmi' into kvmi-v6

Wenzel commented:
@adlazar I used the Web conflict resolution view but I didn't expected to attempt to merge the main `kvmi` branch into `kvmi-v6`.

Let's do this the other way, could you rebase your 3 commits on top of `kvmi` and push `--force` ?

Thanks.

mtarral commented:
The other PR has been merged with your commits.
Closing this one, thank you !
